feat: Add an ability to read metadata for attached databases

Destrolaric opened the pull request:
Hi SQLite team!
Here are my proposed changes to the reading of metadata. To allow to read of metadata from attached databases 
https://www.sqlite.org/lang_attach.html
~~This functionality will be optional and disabled by default because it adds more layers `[schema].[table]` for metadata reading.~~ 
The reason why this functionality will be useful:
It isn't easy to work with attached databases using JDBC API because all metadata reading assumes that we are in the main schema. This is fixable by adding schema information to the existing queries.

gotson commented:
Hi, i don't think we should add a pragma for that. The JDBC methods accept a schema, and if it is `null` then it should not be used for filtering.

Destrolaric commented:
Hi, pragma is used only for getSchemas to mark if we read them at all. Other parts just check for null.

gotson commented:
> Hi, pragma is used only for getSchemas to mark if we read them at all. Other parts just check for null.

i don't think the switch is necessary, we can always retrieve them all

gotson commented on src/main/java/org/sqlite/jdbc3/JDBC3DatabaseMetaData.java:
```diff
@@ -681,7 +681,7 @@ public boolean supportsSavepoints() {
     /** @see java.sql.DatabaseMetaData#supportsSchemasInDataManipulation() */
     public boolean supportsSchemasInDataManipulation() {
-        return false;
+        return true;
```
is this related to the change of this PR, or something we missed ? Database metadata is not DML, thus your changes should not change this value IMHO.

Destrolaric replied:
agree

gotson commented on src/main/java/org/sqlite/jdbc3/JDBC3DatabaseMetaData.java:
```diff
@@ -915,7 +915,7 @@ public ResultSet getColumns(String c, String s, String tblNamePattern, String co
 
         StringBuilder sql = new StringBuilder(700);
-        sql.append("select null as TABLE_CAT, null as TABLE_SCHEM, tblname as TABLE_NAME, ")
+        sql.append("select null as TABLE_CAT, ").append(quote(s)).append(" as TABLE_SCHEM, tblname as TABLE_NAME, ")
```
if `s` is null, then will it quote the `NULL` string thus returning `"NULL"` instead of the SQL `null` ?

Destrolaric replied:
```
    protected static String quote(String tableName) {
        if (tableName == null) {
            return "null";
        } else {
            return String.format("'%s'", tableName);
        }
    }
```
It will be null for SQL requests. And also, it can be seen by unit tests that it will result in a `null` value.

gotson reviewed (commented):
i did a very quick first review

gotson commented:
Code formatting is failing, can you run `mvn spotless:apply` ?

gotson commented on src/main/java/org/sqlite/jdbc3/JDBC3DatabaseMetaData.java:
```diff
@@ -1173,9 +1172,8 @@ public ResultSet getCrossReference(
     /** @see java.sql.DatabaseMetaData#getSchemas() */
```
My understanding is that schemas are currently stored instead of being retrieved everytime, but as databases can be attached dynamically, we should probably get the schemas all the time, and add a test to cover that too.

gotson replied:
still not sure about this one, i see you resolved it, but as `getSchemas` is only initialized if it's null, it would not be updated after a database is attached or detached.

gotson commented on src/test/java/org/sqlite/DBMetaDataTest.java:
```diff
@@ -1398,10 +1400,10 @@ public void columnOrderOfgetImportedKeys() throws SQLException {
         assertThat(importedKeys.next()).isTrue();
         assertThat(importedKeys.getString("PKTABLE_CAT")).isEqualTo("default");
-        assertThat(importedKeys.getString("PKTABLE_SCHEM")).isEqualTo("global");
+        assertThat(importedKeys.getString("PKTABLE_SCHEM")).isEqualTo(null);
```
for this shouldn't we return `main` instead of `null`?

gotson commented on src/test/java/org/sqlite/DBMetaDataTest.java:
```diff
@@ -1423,12 +1425,12 @@ public void columnOrderOfgetExportedKeys() throws SQLException {
         assertThat(exportedKeys.next()).isTrue();
         assertThat(exportedKeys.getString("PKTABLE_CAT")).isEqualTo("default");
-        assertThat(exportedKeys.getString("PKTABLE_SCHEM")).isEqualTo("global");
+        assertThat(exportedKeys.getString("PKTABLE_SCHEM")).isEqualTo(null);
```
same here, it should be the actual schema of the table, not the parameter that was passed

gotson commented on src/test/java/org/sqlite/DBMetaDataTest.java:
```diff
@@ -1423,12 +1425,12 @@ public void columnOrderOfgetExportedKeys() throws SQLException {
+        assertThat(exportedKeys.getString("PKTABLE_SCHEM")).isEqualTo(null);
         assertThat(exportedKeys.getString("FKTABLE_CAT")).isEqualTo("default");
-        assertThat(exportedKeys.getString("FKTABLE_SCHEM")).isEqualTo("global");
+        assertThat(exportedKeys.getString("FKTABLE_SCHEM")).isEqualTo(null);
```
same here, it should be the actual schema of the table, not the parameter that was passed

gotson commented on src/test/java/org/sqlite/DBMetaDataTest.java:
```diff
@@ -1568,4 +1570,313 @@ public void version() throws Exception {
+        }
+
+        @Test
+        public void getTablesForAttachedDatabase() throws SQLException {
```
we should probably have a test where schema is passed as `null` in `getTables`, and verify that returned tables do have the correct `schema` set.

Destrolaric commented:
Hi, I Added 2 test cases for the default DB and for attaching an additional one after `getschemas()` call. Also fixed formatting issues and now `null` schema replaced with `main`

gotson commented on src/main/java/org/sqlite/jdbc3/JDBC3DatabaseMetaData.java:
```diff
@@ -915,7 +915,9 @@ public ResultSet getColumns(String c, String s, String tblNamePattern, String co
         StringBuilder sql = new StringBuilder(700);
-        sql.append("select null as TABLE_CAT, ").append(quote(s)).append(" as TABLE_SCHEM, tblname as TABLE_NAME, ")
+        sql.append("select null as TABLE_CAT, ")
+                .append(quote(s == null ? "main" : s))
+                .append(" as TABLE_SCHEM, tblname as TABLE_NAME, ")
```
I'm not sure about that.

The function should return the actual schema of the table, no? But it seems we just feed back the parameter that was passed.

If we have 2 tables with the same name in 2 different schemas, and we call this function without specifying the schema `s`, then that function should return columns from both tables, but we would need to have `TABLE_SCHEM` to differentiate them.

gotson commented on src/main/java/org/sqlite/jdbc3/JDBC3DatabaseMetaData.java:
```diff
@@ -1198,7 +1206,9 @@ public ResultSet getPrimaryKeys(String c, String s, String table) throws SQLExce
         StringBuilder sql = new StringBuilder(512);
-        sql.append("select null as TABLE_CAT, ").append(quote(s)).append(" as TABLE_SCHEM, '")
+        sql.append("select null as TABLE_CAT, ")
+                .append(quote(s == null ? "main" : s))
```
similar as above, we should return the actual schema of the retrieved PK's table, not the parameter that was passed to the function

gotson commented on src/main/java/org/sqlite/jdbc3/JDBC3DatabaseMetaData.java:
```diff
@@ -1249,7 +1259,7 @@ public ResultSet getExportedKeys(String catalog, String schema, String table)
         catalog = (catalog != null) ? quote(catalog) : null;
 
-        String quotedSchema = (schema != null) ? quote(schema) : null;
+        String quotedSchema = (schema != null) ? quote(schema) : quote("main");
```
similar as above, we should return the actual schema of the retrieved PK's table, not the parameter that was passed to the function

gotson commented on src/main/java/org/sqlite/jdbc3/JDBC3DatabaseMetaData.java:
```diff
@@ -1406,12 +1416,12 @@ public ResultSet getImportedKeys(String catalog, String schema, String table)
                 .append(quote(catalog))
                 .append(" as PKTABLE_CAT, ")
-                .append(quote(schema))
+                .append(quote(schema == null ? "main" : schema))
```
similar as above, we should return the actual schema of the retrieved PK's table, not the parameter that was passed to the function

gotson commented on src/main/java/org/sqlite/jdbc3/JDBC3DatabaseMetaData.java:
```diff
@@ -1406,12 +1416,12 @@ public ResultSet getImportedKeys(String catalog, String schema, String table)
                 .append(quote(catalog))
                 .append(" as FKTABLE_CAT, ")
-                .append(quote(schema))
+                .append(quote(schema == null ? "main" : schema))
```
similar as above, we should return the actual schema of the retrieved PK's table, not the parameter that was passed to the function

gotson commented on src/main/java/org/sqlite/jdbc3/JDBC3DatabaseMetaData.java:
```diff
@@ -1526,7 +1536,7 @@ public ResultSet getIndexInfo(String c, String s, String table, boolean u, boole
         // this is from the JDBC spec, it is part of the driver protocol
         sql.append("select null as TABLE_CAT,")
-                .append(quote(s))
+                .append(quote(s == null ? "main" : s))
```
similar as above, we should return the actual schema of the retrieved PK's table, not the parameter that was passed to the function

gotson commented on src/main/java/org/sqlite/jdbc3/JDBC3DatabaseMetaData.java:
```diff
@@ -1689,7 +1705,10 @@ public synchronized ResultSet getTables(
         sql.append("  NULL AS TABLE_CAT,").append("\n");
-        sql.append("  NULL AS TABLE_SCHEM,").append("\n");
+        sql.append("  ")
+                .append(quote(s == null ? "main" : s))
```
similar as above, we should return the actual schema of the retrieved PK's table, not the parameter that was passed to the function

Destrolaric commented:
@gotson Ok, this one looks doable and will probably, be ready sometime next week.

Destrolaric commented:
@gotson Hi, sorry for waiting; unfortunately, I had no time to finish the pr before. I have a question. JDBC has some functions like getExportedKeys that require using full schema name and returns everything if null. Do we want to read the data from all existing schemas?

Destrolaric commented:
```
        sql.append("    SELECT").append("\n");
        sql.append("      NAME,").append("\n");
        sql.append("      'GLOBAL TEMPORARY' AS TYPE").append("\n");
        sql.append("    FROM").append("\n");
        sql.append("      sqlite_temp_master").append("\n"); 
```
Also question what to do with GLOBAL TEMPORARY tables. They kind of have their scheme now.

Destrolaric commented:
@gotson, Hi, again, can we discuss acceptance criteria regarding this pr? Currently, it needs to add some unit tests, and we can probably discuss if anything else is required. Also it looks like this one leads to the significant change of default behaviour, so it probably may require more accurate handling.

gotson commented:
> @gotson, Hi, again, can we discuss acceptance criteria regarding this pr? Currently, it needs to add some unit tests, and we can probably discuss if anything else is required. Also it looks like this one leads to the significant change of default behaviour, so it probably may require more accurate handling.

Hi, unit tests would be needed yes. What kind of significant changes of default behavior would you foresee though ?

Destrolaric commented:
@gotson Well, there are some. Most importantly, if the value of the schema pattern is null, it will show all DBs instead of only 'main', that may look like a significant API change.

gotson commented:
> @gotson Well, there are some. Most importantly, if the value of the schema pattern is null, it will show all DBs instead of only 'main', that may look like a significant API change.

It's fine, it is more correct

Destrolaric commented:
@gotson I resolved the conflict with the recent two commits

gotson commented:
I still see conflicts though. Are you sure you pushed your latest changes?

Destrolaric commented:
![image](https://user-images.githubusercontent.com/20579256/233098657-c0ae2538-6f22-4d8f-b2e0-29f60ee40fa2.png)
Weird, for me branch have no conflicts

gotson commented:
Weird indeed

![SmartSelect_20230419_220608_Chrome](https://user-images.githubusercontent.com/2139133/233100617-c8927086-902d-44c0-a14e-4372d860f769.jpg)

gotson commented:
My bad, it's only the rebase (which was selected by default) that can't be done.

Squash and merge is ok (and it's what we want).

![SmartSelect_20230419_220755_Chrome](https://user-images.githubusercontent.com/2139133/233101197-aee774b2-cf59-4377-8fbd-45bed2e0e87c.jpg)

gotson commented:
code formatting is failing, can you fix it with `mvn spotless:apply` ?

Destrolaric commented:
@gotson Fixed

gotson commented on src/test/java/org/sqlite/DBMetaDataTest.java:
```diff
@@ -94,7 +94,10 @@ public void getTables() throws SQLException {
         assertThat(rs.getString("TABLE_NAME")).isEqualTo("sqlite_schema");
-        assertThat(rs.next()).isFalse();
+        assertThat(rs.getString("TABLE_SCHEM")).isEqualTo("main");
+        //        assertThat(rs.next()).isTrue();
+        //        assertThat(rs.getString("TABLE_NAME")).isEqualTo("sqlite_schema");
+        //        assertThat(rs.getString("TABLE_SCHEM")).isEqualTo("temp");
```
What's with those commented lines?

gotson commented on src/test/java/org/sqlite/DBMetaDataTest.java:
```diff
@@ -1669,4 +1667,593 @@ public void getPrimaryKeys(String table) throws SQLException {
+        }
+    }
+
+    private void assertSystemSchema(ResultSet rs) throws SQLException {
```
shouldn't that also check for `TABLE_SCHEM` and compare it to a parameter passed in the function?

gotson commented on src/test/java/org/sqlite/DBMetaDataTest.java:
```diff
@@ -417,18 +420,13 @@ public void getColumns() throws SQLException {
+        // SYSTEM TABLE "sqlite_schema" for main
+        assertSystemSchema(rs);
+
+        // SYSTEM TABLE "sqlite_schema" for temp
```
where is that `temp` database attached? i can't find any mention to it

Destrolaric replied:
temp is the sqlite temporary files, I think are returned as system external schema if any exist.

gotson commented on src/test/java/org/sqlite/DBMetaDataTest.java:
```diff
@@ -1001,7 +999,7 @@ public void columnOrderOfgetTables() throws SQLException {
                 meta.getTables(
                         null,
-                        null,
+                        "",
```
I'm not sure about that, i don't think passing `""` should match any schema, that's what passing `null` is for.

gotson commented on src/test/java/org/sqlite/DBMetaDataTest.java:
```diff
@@ -1158,7 +1156,7 @@ public void columnOrderOfgetProcedurColumns() throws SQLException {
     public void columnOrderOfgetSchemas() throws SQLException {
         ResultSet rs = meta.getSchemas();
-        assertThat(rs.next()).isFalse();
+        assertThat(rs.next()).isTrue();
```
shouldn't we have more assertions to verify the content of the result set?

gotson commented on src/test/java/org/sqlite/DBMetaDataTest.java:
```diff
@@ -1669,4 +1667,593 @@ public void getPrimaryKeys(String table) throws SQLException {
+        assertThat(rsmeta.getColumnName(16)).isEqualTo("CHAR_OCTET_LENGTH");
+        assertThat(rsmeta.getColumnName(17)).isEqualTo("ORDINAL_POSITION");
+        assertThat(rsmeta.getColumnName(18)).isEqualTo("IS_NULLABLE");
+        // should be SCOPE_CATALOG, but misspelt in the standard
+        assertThat(rsmeta.getColumnName(19)).isEqualTo("SCOPE_CATLOG");
```
this was fixed in 4429515442f999457e0517216c1e5e0c5b14cb09 and thus should be changed

gotson commented on src/test/java/org/sqlite/DBMetaDataTest.java:
```diff
@@ -1669,4 +1667,593 @@ public void getPrimaryKeys(String table) throws SQLException {
+        }
+
+        @AfterEach
+        public void exit() throws SQLException {
```
it's better for readability to move this function next to the `init()` function

gotson commented on src/test/java/org/sqlite/DBMetaDataTest.java:
```diff
@@ -1669,4 +1667,593 @@ public void getPrimaryKeys(String table) throws SQLException {
+            assertThat(rsMeta.getColumnCount()).isEqualTo(2);
+            assertThat(rsMeta.getColumnName(1)).isEqualTo("TABLE_SCHEM");
+            assertThat(rsMeta.getColumnName(2)).isEqualTo("TABLE_CATALOG");
+            assertThat(rs.getString("TABLE_SCHEM")).isEqualTo("db2");
+            assertThat(rs.next()).isTrue();
+            assertThat(rs.getString("TABLE_SCHEM")).isEqualTo("main");
```
is there some ordering of the schemas mandated by JDBC ? or we do always return them in alphabetical order?

Destrolaric replied:
```
"select name as TABLE_SCHEM, null as TABLE_CATALOG from pragma_database_list order by "
                                    + "TABLE_SCHEM;");
```
Ordering in SQL request

gotson replied:
Just noticed that:

> The results are ordered by TABLE_CATALOG and TABLE_SCHEM

In that case the tests should test for the ordering.

gotson commented on src/test/java/org/sqlite/DBMetaDataTest.java:
```diff
@@ -1669,4 +1667,593 @@ public void getPrimaryKeys(String table) throws SQLException {
+            rs = meta.getTables(null, null, "bob", null);
+            assertThat(rs.next()).isFalse();
+            rs.close();
+            rs = meta.getTables(null, null, "test", null);
+            assertThat(rs.next()).isTrue();
+            assertThat(rs.next()).isFalse();
+            rs.close();
+            rs = meta.getTables(null, null, "test%", null);
+            assertThat(rs.next()).isTrue();
+            assertThat(rs.next()).isTrue();
```
some comments on what is being tested would help

gotson commented on src/test/java/org/sqlite/DBMetaDataTest.java:
```diff
@@ -1669,4 +1667,593 @@ public void getPrimaryKeys(String table) throws SQLException {
+            assertThat(rs.next()).isTrue();
+            rs.close();
+
+            rs = meta.getTables(null, "main", null, new String[] {"table"});
+            assertThat(rs.next()).isTrue();
+            assertThat(rs.getString("TABLE_NAME")).isEqualTo("test");
+            assertThat(rs.next()).isFalse();
+            rs.close();
+
+            rs = meta.getTables(null, "main", null, new String[] {"view"});
+            assertThat(rs.next()).isTrue();
+            assertThat(rs.getString("TABLE_NAME")).isEqualTo("testView");
+            assertThat(rs.next()).isFalse();
+            rs.close();
+
+            rs = meta.getTables(null, "main", null, new String[] {"system table"});
+            assertThat(rs.next()).isTrue();
+            assertThat(rs.getString("TABLE_NAME")).isEqualTo("sqlite_schema");
+            assertThat(rs.next()).isFalse();
+            rs.close();
```
shouldn't we have the same suite of tests for `db2` for consistency?

gotson commented on src/test/java/org/sqlite/DBMetaDataTest.java:
```diff
@@ -1669,4 +1667,593 @@ public void getPrimaryKeys(String table) throws SQLException {
+            stat.executeUpdate("attach database \"" + thirdDB.toURI().toURL() + "\" as db3;");
+            try {
+                schemas = meta.getSchemas();
+                boolean schemaFound = false;
+                while (schemas.next()) {
+                    schemaFound = "db3".equals(schemas.getString("TABLE_SCHEM"));
+                    if (schemaFound) {
+                        break;
+                    }
+                }
+                assertThat(schemaFound).isTrue();
```
maybe put all schemas in a list and verify that it contains all the values we expect, ie `db2`,`db3` and `main`.
If we expect some ordering, we should also test for that here.

gotson commented on src/test/java/org/sqlite/DBMetaDataTest.java:
```diff
@@ -1669,4 +1667,593 @@ public void getPrimaryKeys(String table) throws SQLException {
+        }
+
+        @Test
+        public void testGetSchemasForAttachedDatabases() throws SQLException, IOException {
```
the above function mentions it's about dynamic attachment, but this one doesn't

gotson commented on src/test/java/org/sqlite/DBMetaDataTest.java:
```diff
@@ -1669,4 +1667,593 @@ public void getPrimaryKeys(String table) throws SQLException {
+                assertThat(schemaFound).isTrue();
+
+            } finally {
+                stat.executeUpdate("detach database db3;");
```
should we also test after the detach that the schemas are properly returned?

gotson commented on src/main/java/org/sqlite/jdbc3/JDBC3DatabaseMetaData.java:
```diff
@@ -1175,12 +1229,33 @@ public ResultSet getSchemas() throws SQLException {
+        if (schemaPattern == null) {
+            return getSchemas();
+        }
+        if ("".equals(schemaPattern)) {
+            schemaPattern = "main";
+        }
```
I don't think we should have that behaviour. The documentation says:

> schemaPattern – a schema name; must match the schema name as it is stored in the database; null means schema name should not be used to narrow down the search.

There is no mention of a special handling when passing `"""`

gotson reviewed (changes requested):
For this review i mostly focused on the test class. I think we should have more coverage on dynamic attachment, ie for all functions:
- test the behaviour
- attach 1 db
- test the behaviour
- detach the db
- test the behaviour

Is that something you could add ?
